bt_navigator:
  ros__parameters:
    use_sim_time: False
    global_frame: map
    robot_base_frame: base_link
    odom_topic: /odom
    default_nav_through_poses_bt_xml: "$(find-pkg-share penguin_nav)/config/navigate_through_poses_w_replanning_and_recovery.xml"
    bt_loop_duration: 10
    default_server_timeout: 20
    enable_groot_monitoring: True
    groot_zmq_publisher_port: 1666
    groot_zmq_server_port: 1667
    plugin_lib_names:
    - nav2_compute_path_to_pose_action_bt_node
    - nav2_compute_path_through_poses_action_bt_node
    - nav2_follow_path_action_bt_node
    - nav2_back_up_action_bt_node
    - nav2_spin_action_bt_node
    - nav2_wait_action_bt_node
    - nav2_clear_costmap_service_bt_node
    - nav2_is_stuck_condition_bt_node
    - nav2_goal_reached_condition_bt_node
    - nav2_goal_updated_condition_bt_node
    - nav2_initial_pose_received_condition_bt_node
    - nav2_reinitialize_global_localization_service_bt_node
    - nav2_rate_controller_bt_node
    - nav2_distance_controller_bt_node
    - nav2_speed_controller_bt_node
    - nav2_truncate_path_action_bt_node
    - nav2_goal_updater_node_bt_node
    - nav2_recovery_node_bt_node
    - nav2_pipeline_sequence_bt_node
    - nav2_round_robin_node_bt_node
    - nav2_transform_available_condition_bt_node
    - nav2_time_expired_condition_bt_node
    - nav2_distance_traveled_condition_bt_node
    - nav2_single_trigger_bt_node
    - nav2_is_battery_low_condition_bt_node
    - nav2_navigate_through_poses_action_bt_node
    - nav2_navigate_to_pose_action_bt_node
    - nav2_remove_passed_goals_action_bt_node
    - nav2_planner_selector_bt_node
    - nav2_controller_selector_bt_node
    - nav2_goal_checker_selector_bt_node

bt_navigator_rclcpp_node:
  ros__parameters:
    use_sim_time: False

controller_server:
  ros__parameters:
    use_sim_time: False
    controller_frequency: 5.0
    min_x_velocity_threshold: 0.05
    min_y_velocity_threshold: 0.5
    # min_theta_velocity_threshold: 0.001
    progress_checker_plugin: "progress_checker"
    goal_checker_plugin: "goal_checker"
    controller_plugins: ["FollowPath"]
    odom_topic: "/odom"

    # Progress checker parameters
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.5
      movement_time_allowance: 30.0

    # Goal checker parameters
    goal_checker:
      plugin: "nav2_controller::SimpleGoalChecker"
      #xy_goal_tolerance: 0.2
      xy_goal_tolerance: 0.4
      stateful: True
      #yaw_goal_tolerance: 0.174 # 10 degrees
      yaw_goal_tolerance: 1.57 # 10 degrees
    # RPP Parameters
    FollowPath:
      plugin: "nav2_regulated_pure_pursuit_controller::RegulatedPurePursuitController"
      # --- speed ------------------------------------------------------
      desired_linear_vel: 0.4
      #desired_linear_vel: 0.45
      min_approach_linear_velocity: 0.35
      #min_approach_linear_velocity: 0.45
      approach_velocity_scaling_dist: 1.0
      #approach_velocity_scaling_dist: 0.01
      # --- speed scaling in curvature ------------------------------------------------------
      use_regulated_linear_velocity_scaling: true
      use_cost_regulated_linear_velocity_scaling: true
      regulated_linear_scaling_min_radius: 0.5
      regulated_linear_scaling_min_speed: 0.1
      # --- rotation ------------------------------------------------------
      use_rotate_to_heading: true
      #use_rotate_to_heading: true
      rotate_to_heading_min_angle: 0.5
      #rotate_to_heading_min_angle: 0.349 # 20 degrees
      rotate_to_heading_angular_vel: 0.8
      #rotate_to_heading_angular_vel: 0.5
      #max_angular_accel: 0.5
      max_angular_accel: 1.0
      # --- lookahead point ------------------------------------------------------
      use_velocity_scaled_lookahead_dist: true
      #   --- use_velocity_scaled_lookahead_dist == false
      lookahead_dist: 1.2
      #   --- use_velocity_scaled_lookahead_dist == true
      min_lookahead_dist: 0.8
      max_lookahead_dist: 2.5
      lookahead_time: 1.0
      #   --- velocity scaling in curvature
      use_fixed_curvature_lookahead: true
      curvature_lookahead_dist: 0.1
      # --- reversing ------------------------------------------------------
      allow_reversing: false
      # --- positioning ------------------------------------------------------
      #transform_tolerance: 0.2
      transform_tolerance: 1.57
      # --- collision detection ------------------------------------------------------
      use_collision_detection: true
      max_allowed_time_to_collision: 1.0
      # --- others ------------------------------------------------------
      max_robot_pose_search_dist: 8.0
      use_interpolation: false

controller_server_rclcpp_node:
  ros__parameters:
    use_sim_time: False

local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 5.0
      publish_frequency: 2.0
      global_frame: map
      robot_base_frame: base_link
      use_sim_time: False
      rolling_window: true
      width: 10
      height: 10
      resolution: 0.05
      robot_radius: 0.25
      #robot_radius: 0.4
      plugins: ["obstacle_layer", "inflation_layer"]
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        inflation_radius: 0.3
        #inflation_radius: 0.05
        cost_scaling_factor: 10.0
        #cost_scaling_factor: 1.0
        enabled: true
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        observation_sources: scan
        min_obstacle_height: -1000.0
        scan:
          topic: /scan
          sensor_frame: "base_link" # laserのtfを変更
          max_obstacle_height: 2.0
          min_obstacle_height: -1000.0
          clearing: True
          marking: True
          data_type: "LaserScan"
          raytrace_max_range: 10.0
          raytrace_min_range: 0.1
          obstacle_max_range: 10.0
          obstacle_min_range: 0.1
      always_send_full_costmap: True
  local_costmap_client:
    ros__parameters:
      use_sim_time: False
  local_costmap_rclcpp_node:
    ros__parameters:
      use_sim_time: False

global_costmap:
  global_costmap:
    ros__parameters:
      update_frequency: 1.0
      publish_frequency: 1.0
      global_frame: map
      robot_base_frame: base_link
      width: 250
      height: 250
      origin_x: -50.0
      origin_y: -50.0
      use_sim_time: False
      robot_radius: 0.25
      #robot_radius: 0.4
      resolution: 0.05
      track_unknown_space: true
      plugins: ["obstacle_layer", "inflation_layer"]
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        enabled: True
        cost_scaling_factor: 1.0
        inflation_radius: 0.6
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        observation_sources: scan
        min_obstacle_height: -1000.0
        scan:
          topic: /scan
          sensor_frame: "base_link" # laserのtfを変更
          max_obstacle_height: 2.0
          #min_obstacle_height: -1.0
          min_obstacle_height: -1000.0
          clearing: True
          marking: True
          data_type: "LaserScan"
          raytrace_max_range: 10.0
          raytrace_min_range: 0.1
          obstacle_max_range: 10.0
          obstacle_min_range: 0.1
      always_send_full_costmap: True
  global_costmap_client:
    ros__parameters:
      use_sim_time: False
  global_costmap_rclcpp_node:
    ros__parameters:
      use_sim_time: False

planner_server:
  ros__parameters:
    expected_planner_frequency: 20.0
    use_sim_time: False
    planner_plugins: ["GridBased"]
    GridBased:
      plugin: "nav2_navfn_planner/NavfnPlanner"
      tolerance: 0.5
      use_astar: false
      allow_unknown: true

planner_server_rclcpp_node:
  ros__parameters:
    use_sim_time: False

recoveries_server:
  ros__parameters:
    costmap_topic: local_costmap/costmap_raw
    footprint_topic: local_costmap/published_footprint
    cycle_frequency: 10.0
    recovery_plugins: ["spin", "backup", "wait"]
    spin:
      plugin: "nav2_recoveries/Spin"
    backup:
      plugin: "nav2_recoveries/BackUp"
    wait:
      plugin: "nav2_recoveries/Wait"
    global_frame: map
    robot_base_frame: base_link
    transform_timeout: 0.1
    use_sim_time: False
    simulate_ahead_time: 1.0
    max_rotational_vel: 5.0
    min_rotational_vel: 0.5
    rotational_acc_lim: 0.2

robot_state_publisher:
  ros__parameters:
    use_sim_time: False

waypoint_follower:
  ros__parameters:
    loop_rate: 200
    stop_on_failure: false
    waypoint_task_executor_plugin: "wait_at_waypoint"
    wait_at_waypoint:
      plugin: "nav2_waypoint_follower::WaitAtWaypoint"
      enabled: True
      waypoint_pause_duration: 200
      
velocity_smoother:
  ros__parameters:
    smoothing_frequency: 20.0
    scale_velocities: false
    feedback: "OPEN_LOOP"
    max_velocity: [0.5, 0.0, 0.6]
    min_velocity: [-0.5, 0.0, -0.6]
    #max_velocity: [0.5, 0.0, 2.5]
    #min_velocity: [-0.5, 0.0, -2.5]
    deadband_velocity: [0.0, 0.0, 0.0]
    velocity_timeout: 1.0
    max_accel: [0.2, 0.0, 1.0]
    max_decel: [-0.8, 0.0, -3.2]
    #max_accel: [2.5, 0.0, 3.2]
    #max_decel: [-2.5, 0.0, -3.2]
    odom_topic: "odom"
    odom_duration: 0.1
    use_realtime_priority: false
    enable_stamped_cmd_vel: false

collision_monitor:
  ros__parameters:
    use_sim_time: False
    #base_frame_id: "base_footprint"
    base_frame_id: "base_link"
    odom_frame_id: "map"
    cmd_vel_in_topic: "cmd_vel_smoothed"
    cmd_vel_out_topic: "cmd_vel"
    transform_tolerance: 0.5
    source_timeout: 5.0
    base_shift_correction: True
    stop_pub_timeout: 2.0
    # Polygons represent zone around the robot for "stop" and "slowdown" action types,
    # and robot footprint for "approach" action type.
    # Footprint could be "polygon" type with dynamically set footprint from footprint_topic
    # or "circle" type with static footprint set by radius. "footprint_topic" parameter
    # to be ignored in circular case.
    polygons: ["PolygonStop", "PolygonSlow"]
    PolygonStop:
      type: "polygon"
      #points: [0.5, 0.4, 0.5, -0.4, -0.4, -0.4, -0.4, 0.4]
      points: [0.25, 0.25, 0.25, -0.25, -0.25, -0.25, -0.25, 0.25]
      action_type: "stop"
      max_points: 3
      visualize: True
      polygon_pub_topic: "polygon_stop"
      enabled: True
    PolygonSlow:
      type: "polygon"
      points: [0.7, 0.7, 0.7, -0.7, -0.7, -0.7, -0.7, 0.7]
      action_type: "slowdown"
      max_points: 3
      slowdown_ratio: 0.8
      visualize: True
      polygon_pub_topic: "polygon_slowdown"
      enabled: True
    FootprintApproach:
      type: "polygon"
      action_type: "approach"
      footprint_topic: "/local_costmap/published_footprint"
      time_before_collision: 2.0
      simulation_time_step: 0.1
      max_points: 5
      visualize: False
      enabled: True
    observation_sources: ["scan"]
    scan:
      type: "scan"
      topic: "/scan"
      enabled: True
